src/CorsService.php

```diff
@@ -356,10 +357,65 @@ protected function isOriginAllowed($origin)
         if ($this->isAllOriginsAllowed()) {
             return true;
         }
+        
+        if ($this->hasWildcardOrigins()) {
+            if ($this->isOriginInWildcardOrigins($origin)) {
+                return true;
+            }
+        }
 
         return in_array($origin, $this->allowOrigins);
     }
+	
+	
+    /**
+     * Returns whether or not the origin is allowed as a domain or subdomain of an wildcard origins.
+     *
+     * @param mixed $origin
+     *
+     * @return bool
+     *
+     * @throws InvalidArgument
+     */
+    protected function isOriginInWildcardOrigins($origin)
+    {
+        if (!is_string($origin) || empty($origin)) {
+            throw new InvalidArgument('Origin must be non empty string.');
+        }
 
+        $matchedOrigins = $this->getWildcardOrigins();
+        $hasMatch = false;
+
+        // Let's parse the URI to extract some parts
+        $parser = new UriParser();
+        $originParsed = $parser->parse($origin);
+        foreach ($matchedOrigins as $domain) {
+            // UriParser consider (rightly) *. as an invalid subdomain, so we have to replace it
+            $wildcardDomain = str_replace('*', 'wildcard', $domain);
+            $domainParsed = $parser->parse($wildcardDomain);
+            if ($domainParsed['scheme'] == $originParsed['scheme']) {
+                // Same protocol, next step
+                $domainHost = str_replace('wildcard', '*', $domainParsed['host']);
+                /* 
+                 * Create the final RegExp, like (.*\.)?example\.com
+                 * This would match :
+                 *  - example.com
+                 *  - www.example.com
+                 *  - subdomain.example.com
+                 */
+                // Escape all the dots
+                $domainHostRE = str_replace('.', '\.', $domainHost);
+                // Place the RegExp in the string
+                $domainHostRE = str_replace('*\.', '(.+\.)?', $domainHostRE);
+
+                if (preg_match('#'.$domainHostRE.'#', $originParsed['host']) !== 0) {
```

Comment:
I don’t know if it’s just me, but the logic to get here still seem far more complex than it would have to be. @Jalle19 what do you think?

Reply:
If you need any precision or explanation, I'm around